src/control/cmd/agent/infocache.go

```diff
@@ -61,21 +64,36 @@ type attachInfoCache struct {
 // to choose from.  Returns the index of the device to use.
 func (aic *attachInfoCache) loadBalance(numaNode int) int {
 	aic.mutex.Lock()
+	deviceIndex := invalidIndex
 	numDevs := len(aic.numaDeviceMarshResp[numaNode])
-	deviceIndex := aic.currentNumaDevIdx[numaNode]
-	aic.currentNumaDevIdx[numaNode] = (deviceIndex + 1) % numDevs
+	if numDevs > 0 {
+		deviceIndex = aic.currentNumaDevIdx[numaNode]
+		aic.currentNumaDevIdx[numaNode] = (deviceIndex + 1) % numDevs
+	}
 	aic.mutex.Unlock()
 	return deviceIndex
 }
 
 func (aic *attachInfoCache) getResponse(numaNode int) ([]byte, error) {
 	deviceIndex := aic.loadBalance(numaNode)
+	// If there is no response available for the client's actual NUMA node,
+	// use the default NUMA node
+	if deviceIndex == invalidIndex {
+		deviceIndex = aic.loadBalance(aic.defaultNumaNode)
+		if deviceIndex == invalidIndex {
+			return nil, errors.Errorf("No default response found for the default NUMA node %d", aic.defaultNumaNode)
+		}
+		aic.log.Debugf("No network devices bound to client NUMA node %d.  Using response from NUMA %d", numaNode, aic.defaultNumaNode)
```

Comment:
I would think that this message should also be info-level. The rule of thumb I use for debug vs. info is that info level should be seen in normal operations and alert the user to something relevant. Debug logging is for details that are relevant to developers and/or support folks who need a deeper understanding of what's happening. We run with debug enabled in CI in order to more easily reproduce problems, but real users are expected to run at info level.

The reason I think this should be at info level is that it should be rare, and admins shouldn't have to restart the agent with debug logging enabled in order to see it.

As an aside, it would be nice to have a corresponding check in the libdaos client code to verify that the AttachInfo response received from the agent is for the same numa node as was requested. If not, emitting a message so that the application-side logging picks it up would potentially alert the user to a suboptimal network configuration and they could have that example conversation with their admin team.

Reply:
Yes, making it info level is good.  I was trying to reduce spam, but I agree that it should be rare enough that it doesn't make too much spam.  

With respect to the change in the libdaos client code ... the client doesn't currently make a request for a particular NUMA node.  Originally, I thought we might want to do that so the RPC definition included the NUMA in the request.  But, it was decided to have the agent determine the NUMA node on behalf of the client by examining the client's PID instead.  I could add code to daos_init that detects the client NUMA node with getpid() if we want.  I'd also have to add the detected NUMA node to the dRPC response so the client could compare them.  If you want that particular change, confirm that and lets file a new ticket to handle that.

Reply:
Ah, right. Forgot about that change. I don't think we need to address that right now. Hopefully it should be exceedingly rare for a badly-configured system to make it to the point that actual users are getting suboptimal network configurations.